src/pykeen/nn/node_piece.py

```diff
@@ -626,79 +723,53 @@ def __init__(
                 num_tokens=num_tokens_,
+                token_representation=token_representation,
+                mapped_triples=mapped_triples,
                 num_entities=triples_factory.num_entities,
                 num_relations=triples_factory.real_num_relations,
             )
-            # fill padding (nn.Embedding cannot deal with negative indices)
-            padding = assignment < 0
-            # TODO: different padding idx -> move to tokenizer?
-            assignment[padding] = self.padding_idx = assignment.max().item() + 1
-            assignments.append(assignment)
-
-            # create token representations
-            if isinstance(token_representation, EmbeddingSpecification):
-                token_representation = token_representation.make(
-                    num_embeddings=total_num_tokens,
-                )
-            if token_representation.max_id != total_num_tokens:
-                raise ValueError(
-                    f"If a pre-instantiated representation is provided, it has to have 2 * num_relations + 1= "
-                    f"{total_num_tokens} representations, but has {token_representation.max_id}",
-                )
-            if shape is None:
-                shape = token_representation.shape
-            elif token_representation.shape != tuple(shape):
-                raise ValueError  # TODO
-            
-            token_representations_.append(token_representation)
-        assert shape is not None
+            for tokenizer_inst, token_representation, num_tokens_ in zip(
+                tokenizer_resolver.make_many(queries=tokenizers, kwargs=tokenizers_kwargs),
+                upgrade_to_sequence(token_representations),
+                upgrade_to_sequence(num_tokens),
+            )
+        ]
+
+        # determine shape
+        shape = {t.tokens.shape for t in tokenizations}
+        if len(shape) != 1:
+            raise ValueError(f"Inconsistent token shapes: {shape}")
+        shape = list(shape)[0]
 
         # super init; has to happen *before* any parameter or buffer is assigned
         super().__init__(max_id=triples_factory.num_entities, shape=shape)
 
+        # assign module
+        self.tokenizations = tokenizations
+
         # Assign default aggregation
         self.aggregation = resolve_aggregation(aggregation=aggregation)
-
-        # assign module
-        self.tokens = nn.ModuleList(token_representations_)
-        self.aggregation_index = -(1 + len(token_representation.shape))
-        for i, assignment in enumerate(assignments):
-            self.register_buffer(name=f"assignment_{i}", tensor=assignment)
+        self.aggregation_index = -(1 + len(shape))
 
     def forward(
         self,
         indices: Optional[torch.LongTensor] = None,
     ) -> torch.FloatTensor:  # noqa: D102
-        xs = []
-        for i, tokens in enumerate(self.tokens):
-            # get token IDs, shape: (*, k)
-            token_ids = getattr(self, f"assignment_{i}")
```

Comment:
By moving the tokens + assignment into an own component, we can avoid this ugly attribute lookup.